File: orcastork/adapters/memory/gate.py
"""In-memory ``CooldownGate`` — per-key expiry decided against the injected clock."""

from __future__ import annotations

from ...clock import Clock


class InMemoryCooldownGate:
    def __init__(self, clock: Clock) -> None:
        self._clock = clock
        self._expires_at: dict[str, float] = {}

    async def try_acquire(self, key: str, cooldown_seconds: float) -> bool:
        now = self._clock.monotonic()
        expires_at = self._expires_at.get(key)
        if expires_at is not None and now < expires_at:
            return False
        self._expires_at[key] = now + cooldown_seconds
        return True

File: typescript/src/orcastork/adapters/memory/gate.ts
/**
 * In-memory `CooldownGate` — per-key expiry decided against the injected clock.
 *
 * @module
 */

import type { Clock } from '../../clock.js';
import type { CooldownGate } from '../../ports/cooldown_gate.js';

/** Check-and-arm in one step, with the armed windows held in process memory. */
export class InMemoryCooldownGate implements CooldownGate {
  private readonly clock: Clock;
  private readonly expiresAt = new Map<string, number>();

  public constructor(clock: Clock) {
    this.clock = clock;
  }

  public async tryAcquire(key: string, cooldownMs: number): Promise<boolean> {
    const now = this.clock.monotonic();
    const armedUntil = this.expiresAt.get(key);
    if (armedUntil !== undefined && now < armedUntil) {
      return false;
    }
    this.expiresAt.set(key, now + cooldownMs);
    return true;
  }
}
